File: backend/app/schemas/trace.py
from datetime import datetime
from decimal import Decimal
from typing import Annotated, Self
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    JsonValue,
    StrictInt,
    StringConstraints,
    field_validator,
    model_validator,
)

from app.observability.trace_types import (
    TraceRunType,
    TraceStatus,
    TraceStepType,
)


TraceProviderIdentifier = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=100),
]
TraceModelIdentifier = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=255),
]


class TraceRunCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_type: TraceRunType
    conversation_id: UUID | None = None
    agent_run_id: UUID | None = None
    user_message_id: UUID | None = None
    title: str | None = Field(default=None, max_length=255)
    input_text: str = Field(min_length=1)
    provider: TraceProviderIdentifier | None = None
    model: TraceModelIdentifier | None = None
    status: TraceStatus = TraceStatus.PENDING
    metadata_json: dict[str, JsonValue] = Field(default_factory=dict)

    @field_validator("input_text")
    @classmethod
    def reject_blank_input_text(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("input_text must not be blank")
        return value

    @model_validator(mode="after")
    def require_conversation_for_owned_correlations(self) -> Self:
        if self.conversation_id is None and (
            self.agent_run_id is not None or self.user_message_id is not None
        ):
            raise ValueError(
                "conversation_id is required for agent_run_id or user_message_id"
            )
        return self


class TraceRunRead(TraceRunCreate):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    output_text: str | None
    total_input_tokens: StrictInt | None = Field(default=None, ge=0)
    total_output_tokens: StrictInt | None = Field(default=None, ge=0)
    total_tokens: StrictInt | None = Field(default=None, ge=0)
    estimated_cost: Decimal | None = Field(default=None, ge=0)
    latency_ms: StrictInt | None = Field(default=None, ge=0)
    error_message: str | None
    started_at: datetime | None
    ended_at: datetime | None
    created_at: datetime


class TraceStepCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    trace_run_id: UUID
    step_index: StrictInt = Field(gt=0)
    step_type: TraceStepType
    name: str = Field(min_length=1, max_length=255)
    status: TraceStatus = TraceStatus.PENDING
    input_json: dict[str, JsonValue] = Field(default_factory=dict)

    @field_validator("name")
    @classmethod
    def reject_blank_name(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("name must not be blank")
        return value


class TraceStepRead(TraceStepCreate):
    model_config = ConfigDict(extra="forbid", from_attributes=True)

    id: UUID
    output_json: dict[str, JsonValue] | None
    error_message: str | None
    latency_ms: StrictInt | None = Field(default=None, ge=0)
    started_at: datetime | None
    ended_at: datetime | None
    created_at: datetime

File: frontend/src/types/trace.ts
export type TraceRunType =
  | "chat"
  | "agent"
  | "rag_query"
  | "rag_chat"
  | "evaluation"
  | "tool";

export type TraceStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export type TraceStepType =
  | "build_context"
  | "llm_call"
  | "tool_call"
  | "rag_retrieve"
  | "query_rewrite"
  | "bm25_search"
  | "vector_search"
  | "hybrid_fusion"
  | "parent_child_expand"
  | "rerank"
  | "build_prompt"
  | "final_answer"
  | "eval_metric";

export type RetrievalCandidateSource =
  | "dense"
  | "sparse"
  | "hybrid"
  | "parent"
  | "rerank";

export type TraceRunSummary = {
  id: string;
  run_type: TraceRunType;
  status: TraceStatus;
  title: string | null;
  input_preview: string;
  conversation_id: string | null;
  agent_run_id: string | null;
  user_message_id: string | null;
  provider: string | null;
  model: string | null;
  total_input_tokens: number | null;
  total_output_tokens: number | null;
  total_tokens: number | null;
  estimated_cost: string | null;
  latency_ms: number | null;
  error_message: string | null;
  started_at: string | null;
  ended_at: string | null;
  created_at: string;
};

export type TraceStep = {
  id: string;
  trace_run_id: string;
  step_index: number;
  step_type: TraceStepType;
  name: string;
  status: TraceStatus;
  input_json: Record<string, unknown>;
  output_json: Record<string, unknown> | null;
  error_message: string | null;
  latency_ms: number | null;
  started_at: string | null;
  ended_at: string | null;
  created_at: string;
};

export type RagRetrievalCandidate = {
  id: string;
  retrieval_run_id: string;
  chunk_id: string;
  document_id: string;
  rank: number;
  final_rank: number | null;
  source: RetrievalCandidateSource;
  dense_score: number | null;
  sparse_score: number | null;
  fused_score: number | null;
  rerank_score: number | null;
  selected: boolean;
  content_preview: string;
  metadata_json: Record<string, unknown>;
  created_at: string;
};

export type RagRetrievalRun = {
  id: string;
  trace_run_id: string;
  knowledge_base_id: string;
  strategy_name: "naive_vector";
  original_query: string;
  rewritten_query: string | null;
  top_k: number;
  candidate_count: number;
  selected_count: number;
  score_threshold: number | null;
  latency_ms: number;
  metadata_filter_json: Record<string, unknown>;
  strategy_config_json: Record<string, unknown>;
  created_at: string;
  candidates: RagRetrievalCandidate[];
};

export type TraceRunDetail = Omit<TraceRunSummary, "input_preview"> & {
  input_text: string;
  output_text: string | null;
  metadata_json: Record<string, unknown>;
  steps: TraceStep[];
  retrieval_runs: RagRetrievalRun[];
};
